xapian-core/weight/tfidfweight.cc

```diff
@@ -153,18 +199,29 @@ TfIdfWeight::get_idfn(Xapian::doccount termfreq, char c) const
 	    return (1.0 / termfreq);
 	case 's':
 	    return pow(log(N / termfreq), 2.0);
+	case 'P':
+	    return (log((N+1) / termfreq));
 	default:
 	    AssertEq(c, 't');
 	    return (log(N / termfreq));
     }
 }
 
 double
-TfIdfWeight::get_wtn(double wt, char c) const
+TfIdfWeight::get_wtn(Xapian::termcount doclen, double wt, char c) const
 {
-    (void)c;
-    AssertEq(c, 'n');
-    return wt;
+    switch (c) {
+	case 'P':
+	{
+	    double normlen;
+	    normlen = doclen / get_average_length();
+	    double norm_factor =  1 / (1 - param_slope + (param_slope * normlen));
```

Comment:
Xapian convention is to "cuddle" braces, so that should be:

```
        case 'P': {
```

The rationale is that vertical screen space is limited, and a layout which uses fewer lines without trading off readability means you can view more code in the same height window.

Also it's cleaner to initialise a variable at the point of declaration where you can, like so:

```
    double normlen = doclen / get_average_length();
```